src/uct/ib/dc/dc_mlx5.c

```diff
@@ -1093,6 +1093,50 @@ void uct_dc_mlx5_fc_entry_iter_del(uct_dc_mlx5_iface_t *iface, khiter_t it)
     }
 }
 
+static UCS_F_ALWAYS_INLINE void
+uct_dc_mlx5_fc_restore_wnd(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep,
+                           uct_dc_fc_sender_data_t *fc_data)
+{
+    ucs_arbiter_t *waitq;
+    ucs_arbiter_group_t *group;
+    uint8_t pool_index;
+    int16_t cur_wnd;
+    khiter_t it;
+
+    it = kh_get(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, (uint64_t)ep);
+    if ((fc_data != NULL) &&
+        ((it == kh_end(&iface->tx.fc_hash)) ||
+         (kh_value(&iface->tx.fc_hash, it).seq != fc_data->payload.seq))) {
+        /* Don't process FC_PURE_GRANT further, if an endpoint doesn't exist
+         * anymore, unexpected grant was received on an endpoint or this is not
+         * the same grant sequence number which was expected */
+        return;
+    }
+
+    cur_wnd = ep->fc.fc_wnd;
+
+    uct_rc_fc_restore_wnd(&iface->super.super, &ep->fc);
+    uct_dc_mlx5_fc_entry_iter_del(iface, it);
+
+    if (fc_data == NULL) {
+        /* Restoring FC window not from FC_PURE_GRANT handler, so no other
+         * actions are required */
+        return;
+    }
+
+    UCS_STATS_UPDATE_COUNTER(ep->fc.stats, UCT_RC_FC_STAT_RX_PURE_GRANT, 1);
+
+    /* To preserve ordering we have to dispatch all pending operations if
+     * current fc_wnd is <= 0 */
+    if (cur_wnd <= 0) {
+        uct_dc_mlx5_get_arbiter_params(iface, ep, &waitq, &group,
+                                       &pool_index);
+        ucs_arbiter_group_schedule(waitq, group);
+        uct_dc_mlx5_iface_progress_pending(iface, pool_index);
+        uct_dc_mlx5_iface_check_tx(iface);
+    }
```

Comment:
seems this code should also be only in FC_PURE_GRANT handler

Reply:
done